### PR TITLE
#87 - Reusing storage tests for S3Storage

### PR DESCRIPTION
**Pull request by olegmoz**
Part of issue #87 
Using common storage tests for S3Storage as well, to see it complies common contract for any storage

### CODE REVIEW

**Comment by paulodamaso** [comment]
@g4s8 PLease assign me, I'm available

**Comment by 0crat** [comment]
This pull request [#106](https://github.com/artipie/asto/issues/106) is assigned to @hdouss[/z](https://www.0crat.com/u/hdouss), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/7c0892ce-1fd7-48ba-a4c0-3e63961b3e05); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; we should be aware that @hdouss[/z](https://www.0crat.com/u/hdouss) is on vacation; this ticket may be delayed; there will be a monetary reward for this job

File: src/test/java/com/artipie/asto/StorageExtension.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    public boolean supportsTestTemplate(final ExtensionContext context) {
+        return context.getTestMethod().isPresent();
+    }
+
+    @Override
+    public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContexts(
+        final ExtensionContext context) {
+        final Collection<Storage> storages;
+        try {
+            storages = Arrays.asList(
```
Comment (HDouss) [inline src/test/java/com/artipie/asto/StorageExtension.java]: @olegmoz Please use ListOf

File: src/test/java/com/artipie/asto/StorageExtension.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            .credentialsProvider(
+                StaticCredentialsProvider.create(AwsBasicCredentials.create("foo", "bar"))
+            )
+            .endpointOverride(
+                URI.create(String.format("http://localhost:%d", this.mock.getHttpPort()))
+            )
+            .build();
+        final String bucket = UUID.randomUUID().toString();
+        final S3Storage storage = new S3Storage(client, bucket);
+        client.createBucket(CreateBucketRequest.builder().bucket(bucket).build()).join();
```
Comment (HDouss) [inline src/test/java/com/artipie/asto/StorageExtension.java]: @olegmoz Could we create the bucket before instantiating the storage, and then just return the `S3Storage` constructor result? (so without creating `storage` variable?)

File: src/test/java/com/artipie/asto/StorageExtension.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+         *
+         * @param storage Storage instance to be used in test.
+         */
+        StorageContext(final Storage storage) {
+            this.storage = storage;
+        }
+
+        @Override
+        public String getDisplayName(final int index) {
+            return String.format("[%s]", this.storage.getClass().getSimpleName());
```
Comment (HDouss) [inline src/test/java/com/artipie/asto/StorageExtension.java]: @olegmoz We could use `FormattedText`

File: src/test/java/com/artipie/asto/StorageExtension.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        }
+
+        @Override
+        public List<Extension> getAdditionalExtensions() {
+            return Collections.singletonList(new Resolver());
```
Comment (HDouss) [inline src/test/java/com/artipie/asto/StorageExtension.java]: @olegmoz We could use `ListOf`

File: src/test/java/com/artipie/asto/StorageExtension.java (hunk 2 above)
Comment (olegmoz) [reply src/test/java/com/artipie/asto/StorageExtension.java]: @HDouss thank you, this way it is cleaner, fixed

**Comment by olegmoz** [comment]
@HDouss made changes by one of your comments, do not see point in replacing static methods from Java standard library with Cactoos. Let's see what @g4s8 has to say about it

**Comment by HDouss** [comment]
> @HDouss made changes by one of your comments, do not see point in replacing static methods from Java standard library with Cactoos. Let's see what @g4s8 has to say about it

@olegmoz Good for me! Thanks.

**Comment by 0crat** [comment]
Job audit: `ARC` didn't merge or close PR, performer @hdouss[/z](https://www.0crat.com/u/hdouss)[/z](https://www.0crat.com/u/hdouss) didn't make CR comments

**Comment by 0crat** [comment]
Job was finished in 15 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Pull request merged**